utils.js

```diff
@@ -801,14 +882,117 @@ HDSigner.prototype.signPSBT = async function (psbt) {
+    const input = psbt.txInputs[i]
+    const inputItem = {}
+    inputItem.prev_index = input.index
+    inputItem.prev_hash = input.hash.reverse().toString('hex')
+    if (input.sequence) inputItem.sequence = input.sequence
+    const dataInput = psbt.data.inputs[i]
+    let path = ''
+    if (dataInput.unknownKeyVals && dataInput.unknownKeyVals.length > 1 &&
+      dataInput.unknownKeyVals[1].key.equals(Buffer.from('path')) &&
+      (!dataInput.bip32Derivation || dataInput.bip32Derivation.length === 0)) {
+      path = (dataInput.unknownKeyVals[1].value.toString())
+      inputItem.address_n = convertToAddressNFormat(path)
+    }
+    scriptTypes = psbt.getInputType(i)
+    switch (scriptTypes) {
+      case 'multisig':
+        inputItem.script_type = 'SPENDMULTISIG'
+        break
+      case 'witnesspubkeyhash':
+        inputItem.script_type = 'SPENDWITNESS'
+        break
+      default:
+        inputItem.script_type = isP2WSHScript(psbt.data.inputs[i].witnessUtxo.script) ? 'SPENDP2SHWITNESS' : 'SPENDADDRESS'
+        break
+    }
+    trezortx.inputs.push(inputItem)
+  }
+
+  for (let i = 0; i < psbt.txOutputs.length; i++) {
+    const output = psbt.txOutputs[i]
+    const outputItem = {}
+    const chunks = bjs.script.decompile(output.script)
+    outputItem.amount = output.value.toString()
+    if (chunks[0] === bitcoinops.OP_RETURN) {
+      outputItem.script_type = 'PAYTOOPRETURN'
+      outputItem.op_return_data = chunks[1].toString('hex')
+    } else {
+      if (isBech32(output.address)) {
+        if (output.script.length === 34 &&
+          output.script[0] === 0 &&
+          output.script[1] === 0x20) {
+          outputItem.script_type = 'PAYTOP2SHWITNESS'
+        } else {
+          outputItem.script_type = 'PAYTOWITNESS'
+        }
+      } else {
+        outputItem.script_type = isScriptHash(output.address, this.network) ? 'PAYTOSCRIPTHASH' : 'PAYTOADDRESS'
+      }
+      outputItem.address = output.address
+    }
+    trezortx.outputs.push(outputItem)
+  }
+  return trezortx
+}
+
 /* sign
-Purpose: Create signing information based on HDSigner (if set) and call signPSBT() to actually sign, as well as detect notarization and apply it as required.
+Purpose: Create signing information based on Trezor format
 Param psbt: Required. PSBT object from bitcoinjs-lib
-Returns: psbt from bitcoinjs-lib
+Returns: trezortx or txid
 */
 TrezorSigner.prototype.sign = async function (psbt) {
-  // TODO: sign with trezor connect here
+  if (psbt.txInputs.length <= 0 || psbt.txOutputs.length <= 0 || psbt.version === undefined) {
+    throw new Error('PSBT object is lacking information')
+  }
+  const trezorTx = this.convertToTrezorFormat(psbt)
+  const response = await TrezorConnect.signTransaction(trezorTx)
+  if (response.success === true) {
+    const tx = bjs.Transaction.fromHex(response.payload.serializedTx)
+    for (const i of range(psbt.data.inputs.length)) {
+      const partialSig = [
```

Comment:
this assumes the first witness always, not sure this will work with p2wsh also we need to support legacy non witness signatures which isn't handled here yet